sdk/internal/list/attr.go

```diff
@@ -0,0 +1,251 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package list // import "go.opentelemetry.io/otel/sdk/internal/list"
+
+import "go.opentelemetry.io/otel/attribute"
+
+// Adapted from
+// https://github.com/golang/go/blob/go1.17.6/src/container/list/list.go
+
+// Attribute is an element of a linked list of OpenTelemetry attributes.
+type Attribute struct {
+	// Next and previous pointers in the doubly-linked list of elements.
+	// To simplify the implementation, internally a list l is implemented
+	// as a ring, such that &l.root is both the next element of the last
+	// list element (l.Back()) and the previous element of the first list
+	// element (l.Front()).
+	next, prev *Attribute
+
+	// The list to which this element belongs.
+	list *Attributes
+
+	// The attribute value stored with this element.
+	Value attribute.KeyValue
+}
+
+// Next returns the next list element or nil.
+func (e *Attribute) Next() *Attribute {
+	if p := e.next; e.list != nil && p != &e.list.root {
+		return p
+	}
+	return nil
+}
+
+// Prev returns the previous list element or nil.
+func (e *Attribute) Prev() *Attribute {
+	if p := e.prev; e.list != nil && p != &e.list.root {
+		return p
+	}
+	return nil
+}
+
+// Attributes represents a doubly linked list of OpenTelemetry attributes.
+// The zero value for Attributes is an empty list ready to use.
+type Attributes struct {
+	root Attribute // sentinel list element, only &root, root.prev, and root.next are used
+	len  int       // current list length excluding (this) sentinel element
+}
+
+// Init initializes or clears Attributes a.
+func (a *Attributes) Init() *Attributes {
+	a.root.next = &a.root
+	a.root.prev = &a.root
+	a.len = 0
+	return a
+}
+
+// New returns an initialized list.
+func New() *Attributes { return new(Attributes).Init() }
```

Comment:
```suggestion
func NewAttributes() *Attributes { return new(Attributes).Init() }
```

Given that this would be invoked as `list.New()` it feels insufficiently identifying.